core/src/main/java/hudson/TcpSlaveAgentListener.java

```diff
@@ -194,12 +209,21 @@ public void shutdown() {
          */
         private final int id;
 
-        public ConnectionHandler(Socket s) {
+        public ConnectionHandler(Socket s, ConnectionHandlerFailureCallback parentTerminator) {
             this.s = s;
             synchronized(getClass()) {
                 id = iotaGen++;
             }
             setName("TCP agent connection handler #"+id+" with "+s.getRemoteSocketAddress());
+            setUncaughtExceptionHandler((t, e) -> {
+                LOGGER.log(Level.SEVERE, "Uncaught exception in TcpSlaveAgentListener ConnectionHandler " + t, e);
+                try {
+                    parentTerminator.run(e);
+                    s.close();
```

Comment:
It may cay a race condition, because the current socket may be still handling Socket when the parent thread restarts. It may happen if the Thread is delayed for 5 seconds, which is probably non-realistic.

Reply:
Maybe there should be different order

Reply:
Flipped the order.